torch/profiler/profiler.py

```diff
@@ -0,0 +1,191 @@
+        STOP_TRACE = 2
+
+    class State(object):
+        INACTIVE = 0
+        WARMUP = 1
+        ACTIVE = 2
+
+    def __init__(self, wait: int, warmup: int, active: int, output_fn: Optional[Callable[[prof.profile], None]]):
+        assert wait >= 0 and warmup >= 0 and active > 0
+        if warmup == 0:
+            print("Warning: profiler won't be using a warmup, which can skew profiler results")
+        self.wait = wait
+        self.warmup = warmup
+        self.active = active
+        self.output_fn = output_fn
+
+        def active_active_fn(step):
+            if self._mod_step(step) == 1:
+                return [EnablePred.Action.STOP_TRACE, EnablePred.Action.START_WARMUP, EnablePred.Action.START_TRACE]
+            else:
+                return []
+
+        def inactive_warmup_fn(_):
+            raise RuntimeError("Incorrect profiler state sequence")
+
+        self.actions_map = {
+            EnablePred.State.ACTIVE: {
+                EnablePred.State.ACTIVE: active_active_fn,
+                EnablePred.State.WARMUP: [EnablePred.Action.START_TRACE],
+                EnablePred.State.INACTIVE: [EnablePred.Action.START_WARMUP, EnablePred.Action.START_TRACE],
+            },
+            EnablePred.State.WARMUP: {
+                EnablePred.State.ACTIVE: [EnablePred.Action.STOP_TRACE, EnablePred.Action.START_WARMUP],
+                EnablePred.State.WARMUP: [],
+                EnablePred.State.INACTIVE: [EnablePred.Action.START_WARMUP],
+            },
+            EnablePred.State.INACTIVE: {
+                EnablePred.State.ACTIVE: [EnablePred.Action.STOP_TRACE],
+                EnablePred.State.WARMUP: inactive_warmup_fn,
+                EnablePred.State.INACTIVE: [],
+            }
+        }
+
+    def _mod_step(self, step: int):
+        sum_states = self.wait + self.warmup + self.active
+        r = step % sum_states
+        if r == 0:
+            r = sum_states
+        return r
+
+    def _num_state(self, step: int):
+        mod_step = self._mod_step(step)
+        if mod_step <= self.wait:
+            return EnablePred.State.INACTIVE
+        elif mod_step <= self.wait + self.warmup:
+            return EnablePred.State.WARMUP
+        else:
+            return EnablePred.State.ACTIVE
+
+    def actions(self, step: int):
+        if step == 1:
+            st = self._num_state(step)
+            if st == EnablePred.State.ACTIVE:
+                return [EnablePred.Action.START_WARMUP, EnablePred.Action.START_TRACE]
+            elif st == EnablePred.State.WARMUP:
+                return [EnablePred.Action.START_WARMUP]
+            else:
+                return []
+        else:
+            st = self._num_state(step)
+            prev_st = self._num_state(step - 1)
+            acts = self.actions_map[st][prev_st]
+            if callable(acts):
+                return acts(step)
+            else:
+                return acts
+
+
+class profile(object):
+    """
+    PyTorch profiler context manager.
```

Comment:
Elaborate here on what the profiler can provide and how it's intended to be used. People should have an idea of what this looks like before reading the example. Mention the critical functions, how to use them, and why they're important.

Reply:
Instead of mentioning the critical functions in this preamble, this could add docstrings to the methods.